internal/naming/limitations.md

```diff
@@ -0,0 +1,114 @@
+
+Label values must be 63 characters or less and can be empty.
+
+Both label names and values must be alphanumeric with hyphens U+002D, underscores U+005F,
+and dots U+002E allowed in between.
+
+- [k8s.io/apimachinerypkg/util/validation.IsQualifiedName](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/validation#IsQualifiedName)
+- [k8s.io/apimachinerypkg/util/validation.IsValidLabelValue](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/validation#IsValidLabelValue)
+
+
+# Annotations
+
+[k8s-annotations]: https://docs.k8s.io/concepts/overview/working-with-objects/annotations/
+
+Annotation names must be 317 characters or less. The portion before an optional slash U+002F
+must be a DNS subdomain. The portion after must be 63 characters or less and alphanumeric with
+hyphens U+002D, underscores U+005F, and dots U+002E allowed in between.
+
+Annotation values may contain anything, but the combined size of *all* names and values
+must be 256 KiB or less.
+
+- [https://pkg.go.dev/k8s.io/apimachinery/pkg/api/validation.ValidateAnnotations](https://pkg.go.dev/k8s.io/apimachinery/pkg/api/validation#ValidateAnnotations)
+
+
+# Specifics
+
+The Kubernetes API validates custom resource metadata.
+[Custom resource names are DNS subdomains](https://releases.k8s.io/v1.23.0/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/validator.go#L60).
+It may be possible to limit this further through validation. This is a stated
+goal of [CEL expression validation](https://docs.k8s.io/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules).
+
+[ConfigMap names are DNS subdomains](https://releases.k8s.io/v1.23.0/pkg/apis/core/validation/validation.go#L5618).
+
+[CronJob names are DNS subdomains](https://docs.k8s.io/concepts/workloads/controllers/cron-jobs/)
+but must be [52 characters or less](https://releases.k8s.io/v1.23.0/pkg/apis/batch/validation/validation.go#L281).
+
+[Deployment names are DNS subdomains](https://releases.k8s.io/v1.23.0/pkg/apis/apps/validation/validation.go#L632).
+
+[Job names are DNS subdomains](https://releases.k8s.io/v1.23.0/pkg/apis/batch/validation/validation.go#L86).
+When `.spec.completionMode = Indexed`, the name must be shorter (closer to 61 characters, it depends).
+When `.spec.manualSelector` is unset, its Pods get (and must have) a "job-name" label, limiting the
+name to 63 characters or less.
+
+[Namespace names are DNS labels](https://releases.k8s.io/v1.23.0/pkg/apis/core/validation/validation.go#L5963).
+
+[PersistentVolumeClaim (PVC) names are DNS subdomains](https://releases.k8s.io/v1.23.0/pkg/apis/core/validation/validation.go#L2066).
+
+[Pod names are DNS subdomains](https://releases.k8s.io/v1.23.0/pkg/apis/core/validation/validation.go#L3443).
+The strategy for [generating Pod names](https://releases.k8s.io/v1.23.0/pkg/registry/core/pod/strategy.go#L62) truncates to 63 characters.
+The `.spec.hostname` field must be 63 characters or less.
+
+PodDisruptionBudget (PDB)
```

Comment:
:+1: I like having these here for tracking purposes even if we don't yet have them filled in yet.